pkg/controllers/mpi_job_controller.go

```diff
@@ -0,0 +1,997 @@
+		if config, err := c.getConfigMap(mpiJob, workerReplicas, gpusPerWorker); config == nil || err != nil {
+			return err
+		}
+
+		// Get the launcher ServiceAccount for this MPIJob.
+		if sa, err := c.getLauncherServiceAccount(mpiJob); sa == nil || err != nil {
+			return err
+		}
+
+		// Get the launcher Role  for this MPIJob.
+		if r, err := c.getLauncherRole(mpiJob, workerReplicas); r == nil || err != nil {
+			return err
+		}
+
+		// Get the launcher RoleBinding for this MPIJob.
+		if rb, err := c.getLauncherRoleBinding(mpiJob); rb == nil || err != nil {
+			return err
+		}
+	}
+
+	worker, err := c.getWorkerStatefulSet(mpiJob, workerReplicas)
+	if err != nil {
+		return err
+	}
+
+	// If the worker is ready, start the launcher.
+	workerReady := workerReplicas == 0 || int(worker.Status.ReadyReplicas) == workerReplicas
+	if workerReady && launcher == nil {
+		launcherGPUs := totalGPUs
+		if launcherGPUs > c.gpusPerNode {
+			launcherGPUs = c.gpusPerNode
+		}
+		launcher, err = c.kubeClient.BatchV1().Jobs(namespace).Create(newLauncher(mpiJob, launcherGPUs, c.kubectlDeliveryImage))
+		if err != nil {
+			return err
+		}
+	}
+
+	// Finally, we update the status block of the MPIJob resource to reflect the
+	// current state of the world.
+	err = c.updateMPIJobStatus(mpiJob, launcher, worker)
+	if err != nil {
+		return err
+	}
+
+	c.recorder.Event(mpiJob, corev1.EventTypeNormal, SuccessSynced, MessageResourceSynced)
+	return nil
+}
+
+// getLauncherJob gets the launcher Job controlled by this MPIJob.
+func (c *MPIJobController) getLauncherJob(mpiJob *kubeflow.MPIJob) (*batchv1.Job, error) {
+	launcher, err := c.jobLister.Jobs(mpiJob.Namespace).Get(mpiJob.Name + launcherSuffix)
+	if errors.IsNotFound(err) {
+		return nil, nil
+	}
+	if err != nil {
+		// If an error occurs during Get, we'll requeue the item so we can
+		// attempt processing again later. This could have been caused by a
+		// temporary network failure, or any other transient reason.
+		return nil, err
+	}
+
+	// If the launcher is not controlled by this MPIJob resource, we should log
+	// a warning to the event recorder and return.
+	if !metav1.IsControlledBy(launcher, mpiJob) {
+		msg := fmt.Sprintf(MessageResourceExists, launcher.Name)
+		c.recorder.Event(mpiJob, corev1.EventTypeWarning, ErrResourceExists, msg)
+		return launcher, fmt.Errorf(msg)
+	}
+
+	return launcher, nil
+}
+
+// getTotalGPUs gets the total number of desired GPUs. Defaults to 1 if not specified.
+func getTotalGPUs(mpiJob *kubeflow.MPIJob) int {
+	totalGPUs := 1
+	if mpiJob.Spec.GPUs != nil {
+		totalGPUs = int(*mpiJob.Spec.GPUs)
+	}
+	return totalGPUs
+}
+
+// getWorkerReplicas gets the desired number of worker replicas.
+func (c *MPIJobController) getWorkerReplicas(totalGPUs int, done bool) int {
+	workerReplicas := 0
+	if totalGPUs > c.gpusPerNode {
+		// The launcher also does work, so the # worker replicas needed is deducted by 1.
+		workerReplicas = totalGPUs/c.gpusPerNode - 1
```

Comment:
What happens if you request 6 GPUs, and gpusPerNode = 4?

Reply:
That should be an error. https://github.com/kubeflow/mpi-operator/issues/3